tests/trainer/test_trainer.py

```diff
@@ -1,247 +1,375 @@
+            'model': SimpleModel(),
+            'train_dataloader': DataLoader(
+                dataset=RandomClassificationDataset(),
+                batch_size=4,
+                shuffle=True,
+            ),
+            'eval_dataloader': DataLoader(
+                dataset=RandomClassificationDataset(),
+                shuffle=False,
+            ),
+            'max_duration': '2ep',
+            'seed': 0,
+            'device': device,
+            'precision': precision,
+            'deterministic_mode': True,  # testing equivalence
+            'loggers': [],  # no progress bar
+        }
+
+    @pytest.fixture(autouse=True)
+    def create_reference_model(self, config, tmpdir_factory, *args):
+        """Trains the reference model, and saves checkpoints."""
+        save_folder = tmpdir_factory.mktemp("{device}-{precision}".format(**config))
+        config.update({'save_interval': '1ep', 'save_folder': save_folder})
+
+        trainer = Trainer(**config)
+        trainer.fit()
+
+        self.reference_model = trainer.state.model
+        self.reference_folder = save_folder
+
+    def test_determinism(self, config, *args):
+        trainer = Trainer(**config)
+        trainer.fit()
+
+        self.assert_models_equal(trainer.state.model, self.reference_model)
+
+    def test_grad_accum(self, config, *args):
+        config.update({
+            'grad_accum': 2,
+        })
+
+        trainer = Trainer(**config)
+        trainer.fit()
+
+        self.assert_models_equal(trainer.state.model, self.reference_model)
+
+    def test_checkpoint(self, config, *args):
+        # load from epoch 1 checkpoint and finish training
+        checkpoint_file = os.path.join(self.reference_folder, 'ep1.tar')
+        config['load_path'] = checkpoint_file
+
+        trainer = Trainer(**config)
+        assert trainer.state.timer.epoch == "1ep"  # ensure checkpoint state loaded
+        trainer.fit()
+
+        self.assert_models_equal(trainer.state.model, self.reference_model)
+
+    def test_model_init(self, config, *args):
+        # as a control test, we reinitialize the model weights, and
+        # expect the resulting trained model to differe from the reference.
+        config['model'] = SimpleModel()
+
+        trainer = Trainer(**config)
+        trainer.fit()
+
+        with pytest.raises(AssertionError):
+            self.assert_models_equal(trainer.state.model, self.reference_model)
+
+
+class AssertDataAugmented(Callback):
+    """Helper callback that asserts test whether the augmented batch was passed to the model during the forward pass.
+    The original batch is passed through the model and we assert that the outputs are not the same.
+
+    Assumes batch_size=1, and no gradient accumulation.
+    """
+
+    def __init__(self, dataset):
+        self.dataset = dataset
+
+    def after_forward(self, state, logger):
+        batch_idx = state.timer.batch_in_epoch.value
+        batch_size = state.train_dataloader.batch_size
```

Comment:
state.batch_num_samples
```suggestion
        batch_size = state.batch_num_samples
```